src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs

```diff
@@ -0,0 +1,386 @@
+                WHERE timeline = $1
+            RETURNING write_ts;
+        "#;
+        let client = self
+            .get_connection()
+            .await
+            .expect("todo: add internal retries");
+        let statement = client
+            .prepare_cached(q)
+            .await
+            .expect("todo: add internal retries");
+        let result = client
+            .query_one(&statement, &[&self.timeline, &proposed_next_ts])
+            .await
+            .expect("todo: add internal retries");
+
+        let write_ts: i64 = result.try_get("write_ts").expect("missing column write_ts");
+        let write_ts = Timestamp::try_from(write_ts).expect("we don't use negative timestamps");
+
+        debug!(write_ts = ?write_ts, "write_ts");
+
+        let advance_to = write_ts.step_forward();
+
+        WriteTimestamp {
+            timestamp: write_ts,
+            advance_to,
+        }
+    }
+
+    #[tracing::instrument(name = "oracle::peek_write_ts", level = "debug", skip_all)]
+    async fn peek_write_ts(&self) -> Timestamp {
+        let q = r#"
+            SELECT write_ts FROM timestamp_oracle
+                WHERE timeline = $1;
+        "#;
+        let client = self
+            .get_connection()
+            .await
+            .expect("todo: add internal retries");
+        let statement = client
+            .prepare_cached(q)
+            .await
+            .expect("todo: add internal retries");
+        let result = client
+            .query_one(&statement, &[&self.timeline])
+            .await
+            .expect("todo: add internal retries");
+
+        let write_ts: i64 = result.try_get("write_ts").expect("missing column write_ts");
+        let write_ts = Timestamp::try_from(write_ts).expect("we don't use negative timestamps");
+
+        debug!(write_ts = ?write_ts, "peek_write_ts");
+
+        write_ts
+    }
+
+    #[tracing::instrument(name = "oracle::read_ts", level = "debug", skip_all)]
+    async fn read_ts(&self) -> Timestamp {
+        let q = r#"
+            SELECT read_ts FROM timestamp_oracle
+                WHERE timeline = $1;
+        "#;
+        let client = self
+            .get_connection()
+            .await
+            .expect("todo: add internal retries");
+        let statement = client
+            .prepare_cached(q)
+            .await
+            .expect("todo: add internal retries");
+        let result = client
+            .query_one(&statement, &[&self.timeline])
+            .await
+            .expect("todo: add internal retries");
+
+        let read_ts: i64 = result.try_get("read_ts").expect("missing column read_ts");
+        let read_ts = Timestamp::try_from(read_ts).expect("we don't use negative timestamps");
+
+        debug!(read_ts = ?read_ts, "read_ts");
+
+        read_ts
+    }
+
+    #[tracing::instrument(name = "oracle::apply_write", level = "debug", skip_all)]
+    async fn apply_write(&mut self, write_ts: Timestamp) {
+        let q = r#"
```

Comment:
The query from the design document for `apply_write` is 

````
UPDATE
    timestamp_oracle SET write_ts = GREATEST(write_ts, $write_ts),
    read_ts = GREATEST(read_ts, write_ts)
    WHERE timeline = $timeline;
```

I think the query from the design document is suspect, `write_ts` that is referenced in ` read_ts = GREATEST(read_ts, write_ts)` will actually be the old `write_ts` value, the one from before the ` write_ts = GREATEST(write_ts, $write_ts)` part of the update.

Reply:
I am not qualified unfortunately to say as I do not know enough about timestamp advancement. I just wanted to flag the discrepancy.

Reply:
Calling @jkosh44, I have far less context on specifically the guarantee we need here.

It's possible that neither of them are correct if the old read_ts and the new write_ts are both less than the old write_ts? E.g. if the old read_ts is 1, the new write_ts is 2, and the old write_ts is 3, then we'd leave the write_ts at 3 and only advance the read_ts to 2. Is that what we want?

Regardless of the outcome, I think this file wants some unit tests a la the persist ones that are gated behind an env var with a postgres url

Reply:
Oooh, we should copy Joe's 3 properties and 3 scenarios/outcomes in as a comment :)